---
title: "Bin Stats Exploration"
author: "Eric Zhao"
date: "4/23/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.path = 'figures/', dev = c('png', 'pdf'), dpi = 300)

library(GenomicRanges)
library(tidyverse)
library(hexbin)
library(viridis)
library(cowplot)
library(pscl)
library(MASS)
library(annotatr)
library(flexmix)
library(countreg)
library(modelr)
library(diptest)
library(rstan)
```

```{r cpg_annotations}
annots = c('hg38_cpgs')
annotations = build_annotations(genome = 'hg38', annotations = annots)

```

```{r custom_functions}
plot_scatter_bin_methylation_linear <- function(bin_methylation_data, nbfit) {
  scatterplot_data <- bin_methylation_data %>%
    group_by(gc_bin) %>%
    summarise(max_cpg_count = max(cpg_count)) %>%
    ungroup() %>%
    filter(max_cpg_count > 0) %>%
    expand_grid(tibble(cpg_count = 1:50)) %>%
    filter(cpg_count <= max_cpg_count) %>%
    mutate(
      gc_content = gc_bin %>% as.character %>% as.numeric
    ) %>%
    mutate(
      predicted_coverage = predict(nbfit, newdata=.)
    )

  bin_methylation_subset %>%
    ggplot(aes(
      x = cpg_count,
      y = coverage_int
    )) +
    geom_point(alpha=0.1, shape=20) +
    geom_line(aes(
      x = cpg_count,
      y = predicted_coverage
    ), color = 'red', data = scatterplot_data) +
    facet_wrap(~ gc_bin, scales='free') +
    ylim(0, 100) +
    labs(
      x = 'CpG count',
      y = 'Coverage'
    )
}
plot_scatter_bin_methylation <- function(bin_methylation_data, nbfit) {
  scatterplot_data <- bin_methylation_data %>%
    group_by(gc_bin) %>%
    summarise(max_cpg_count = max(cpg_count)) %>%
    ungroup() %>%
    filter(max_cpg_count > 0) %>%
    expand_grid(tibble(cpg_count = 1:50)) %>%
    filter(cpg_count <= max_cpg_count) %>%
    mutate(
      gc_content = gc_bin %>% as.character %>% as.numeric
    ) %>%
    mutate(
      predicted_coverage = exp(predict(nbfit, newdata=.))
    )

  bin_methylation_subset %>%
    ggplot(aes(
      x = cpg_count,
      y = coverage_int
    )) +
    geom_point(alpha=0.1, shape=20) +
    geom_line(aes(
      x = cpg_count,
      y = predicted_coverage
    ), color = 'red', data = scatterplot_data) +
    facet_wrap(~ gc_bin, scales='free') +
    ylim(0, 100) +
    labs(
      x = 'CpG count',
      y = 'Coverage'
    )
}

plot_bin_methylation_fit <- function(bin_methylation_data, nbfit) {
  bin_methylation_fit_plot_data_raw <- bin_methylation_data %>%
    group_by(cpg_bin, gc_bin) %>%
    summarise(count = n()) %>%
    ungroup() %>%
    expand_grid(tibble(x = 1:50)) %>%
    mutate(
      gc_content = gc_bin %>% as.character %>% as.numeric,
      cpg_count = cpg_bin %>% as.character %>% as.integer
    )
  
  bin_methylation_fit_plot_data <- bin_methylation_fit_plot_data_raw %>%
    mutate(
      unmethylated_mu = exp(predict(zero_mu_fit, newdata=.)),
    unmethylated_theta = exp(predict(zero_theta_fit, newdata=.)),
      unmethylated_fit = dnbinom(x, mu = unmethylated_mu, size = unmethylated_theta) * count,
      methylated_mu = ifelse(cpg_count == 0, NA, exp(predict(nbfit, newdata = . ))),
      methylated_theta = nbfit$theta,
      methylated_fit = dnbinom(x, mu = methylated_mu, size =  methylated_theta) * count
    ) %>%
    filter(!is.nan(methylated_fit))
    
  bin_methylation_fit_plot_data_reshaped <- bin_methylation_fit_plot_data %>%
    dplyr::select(cpg_bin, gc_bin, x, methylated_fit, unmethylated_fit) %>%
    gather(fit, value, -cpg_bin, -gc_bin, -x) %>%
    arrange(cpg_bin, gc_bin) %>%
    mutate(facet_label = sprintf(
        '%s CpGs, GC: %s%%',
        cpg_bin %>% as.character,
        gc_bin %>% as.character %>% as.numeric * 100 %>% round
      )) %>%
    mutate(facet_label = factor(facet_label, levels = unique(facet_label)))
  
  return(
    bin_methylation_data %>%
      arrange(cpg_bin, gc_bin) %>%
      mutate(facet_label = sprintf(
        '%s CpGs, GC: %s%%',
        cpg_bin %>% as.character,
        gc_bin %>% as.character %>% as.numeric * 100 %>% round
      )) %>%
      mutate(facet_label = factor(facet_label, levels = unique(facet_label))) %>%
      ggplot(aes(
        x = coverage_int,
      )) +
      geom_histogram(aes(
        fill = methylation_status
      ), alpha=0.5) +
      geom_line(aes(
        x = x,
        y = value,
        group = fit,
        color = fit
      ), data = bin_methylation_fit_plot_data_reshaped) +
      facet_wrap(~ facet_label, scales = 'free') +
      labs(
        x = 'Coverage',
        y = 'Number of bins',
        color = 'Model fit',
        fill = 'Classification'
      )
  )
}

plot_bin_methylation_fit_linear <- function(bin_methylation_data, nbfit) {
  bin_methylation_fit_plot_data_raw <- bin_methylation_data %>%
    group_by(cpg_bin, gc_bin) %>%
    summarise(count = n()) %>%
    ungroup() %>%
    expand_grid(tibble(x = 1:50)) %>%
    mutate(
      gc_content = gc_bin %>% as.character %>% as.numeric,
      cpg_count = cpg_bin %>% as.character %>% as.integer
    )
  
  bin_methylation_fit_plot_data <- bin_methylation_fit_plot_data_raw %>%
    mutate(
      unmethylated_mu = exp(predict(zero_mu_fit, newdata=.)),
    unmethylated_theta = exp(predict(zero_theta_fit, newdata=.)),
      unmethylated_fit = dnbinom(x, mu = unmethylated_mu, size = unmethylated_theta) * count,
      methylated_mu = ifelse(cpg_count == 0, NA, predict(nbfit, newdata = . )),
      methylated_theta = nbfit$theta,
      methylated_fit = dnbinom(x, mu = methylated_mu, size =  methylated_theta) * count
    ) %>%
    filter(!is.nan(methylated_fit))
    
  bin_methylation_fit_plot_data_reshaped <- bin_methylation_fit_plot_data %>%
    dplyr::select(cpg_bin, gc_bin, x, methylated_fit, unmethylated_fit) %>%
    gather(fit, value, -cpg_bin, -gc_bin, -x) %>%
    arrange(cpg_bin, gc_bin) %>%
    mutate(facet_label = sprintf(
        '%s CpGs, GC: %s%%',
        cpg_bin %>% as.character,
        gc_bin %>% as.character %>% as.numeric * 100 %>% round
      )) %>%
    mutate(facet_label = factor(facet_label, levels = unique(facet_label)))
  
  return(
    bin_methylation_data %>%
      arrange(cpg_bin, gc_bin) %>%
      mutate(facet_label = sprintf(
        '%s CpGs, GC: %s%%',
        cpg_bin %>% as.character,
        gc_bin %>% as.character %>% as.numeric * 100 %>% round
      )) %>%
      mutate(facet_label = factor(facet_label, levels = unique(facet_label))) %>%
      ggplot(aes(
        x = coverage_int,
      )) +
      geom_histogram(aes(
        fill = methylation_status
      ), alpha=0.5) +
      geom_line(aes(
        x = x,
        y = value,
        group = fit,
        color = fit
      ), data = bin_methylation_fit_plot_data_reshaped) +
      facet_wrap(~ facet_label, scales = 'free') +
      labs(
        x = 'Coverage',
        y = 'Number of bins',
        color = 'Model fit',
        fill = 'Classification'
      )
  )
}
```

```{r human_results}
chr <- read_tsv('~/../Downloads/by_chromosome/bin_stats_human_chr3.tsv', comment = '#') %>%
  mutate(coverage_int = mean_coverage %>% round %>% as.integer) %>%
  filter(mean_coverage > 0 | gc_content > 0 | cpg_count > 0)

arab1 <- read_tsv('~/../Downloads/by_chromosome/bin_stats_arabidopsis_Arabidopsis1.tsv', comment = '#') %>%
  mutate(coverage_int = mean_coverage %>% round %>% as.integer) %>%
  filter(mean_coverage > 0 | gc_content > 0 | cpg_count > 0)

arab3 <- read_tsv('~/../Downloads/by_chromosome/bin_stats_arabidopsis_Arabidopsis1.tsv', comment = '#') %>%
  mutate(coverage_int = mean_coverage %>% round %>% as.integer) %>%
  filter(mean_coverage > 0 | gc_content > 0 | cpg_count > 0)

annotations_cpg <- chr %>%
  dplyr::select(
    chr = bin_chr,
    start = bin_start,
    end = bin_end
  ) %>%
  GRanges() %>%
  annotate_regions(
    regions = .,
    annotations = annotations,
    ignore.strand = TRUE,
    quiet = FALSE
  ) %>%
  as_tibble() %>%
  mutate(
    cpg_region = factor(gsub('hg38_cpg_', '', annot.type), levels = c('inter', 'shelves', 'shores', 'islands'))
  ) %>%
  group_by(seqnames, start, end) %>%
  filter(as.integer(cpg_region) == max(as.integer(cpg_region))) %>%
  ungroup()

chr_cpg_annotated <- chr %>%
  left_join(
    annotations_cpg %>%
      dplyr::select(
        bin_chr = seqnames,
        bin_start = start,
        bin_end = end,
        cpg_region
      )
  ) %>%
  mutate(
    cpg_bin = factor(round(round(cpg_count * 20 / max(cpg_count)) * max(cpg_count) / 20)),
    gc_bin = factor(round(gc_content * 4, 1) / 4)
  )
```

```{r heatmap_coverage_by_cpg_gc, fig.height=5, fig.width = 6}
chr_cpg_annotated %>%
  filter(!is.na(cpg_region)) %>%
  mutate(gc_content = round(gc_content, 2)) %>%
  group_by(cpg_count, gc_content, cpg_region) %>%
  summarise(mean_mean_coverage = mean(mean_coverage)) %>%
  ungroup() %>%
  ggplot(aes(
    x = gc_content,
    y = cpg_count,
    fill = log(mean_mean_coverage)
  )) +
  geom_tile() +
  scale_fill_viridis() +
  facet_wrap(cpg_region ~ .) +
  labs(
    x = 'GC content',
    y = 'CpG per 300 bp bin',
    fill = 'Log mean coverage'
  )
```

```{r modeling_coverge_by_cpg_count, fig.width = 5, fig.height = 5}
coverage_by_cpg_count <- chr_cpg_annotated %>%
  group_by(cpg_count) %>%
  summarise(mean_coverage = mean(mean_coverage), count = n())

coverage_by_cpg_count_model <- lm(
  log(mean_coverage) ~ offset(log(cpg_count)) + cpg_count,
  data = coverage_by_cpg_count %>% filter(cpg_count >= 12, mean_coverage != 0)
)

coverage_by_cpg_count_model$coefficients

fitted_data <- tibble(
    cpg_count = seq(min(coverage_by_cpg_count$cpg_count), max(coverage_by_cpg_count$cpg_count))
  ) %>%
  mutate(
    fitted_coverage = if_else(cpg_count == 0, 0, exp(predict(coverage_by_cpg_count_model, newdata = .))),
    fitted_manual = exp(coverage_by_cpg_count_model$coefficients[[1]]) * cpg_count * exp(coverage_by_cpg_count_model$coefficients[[2]] * cpg_count)
  )

chr_cpg_annotated %>%
  ggplot(aes(
    x = cpg_count,
    y = mean_coverage
  )) +
  geom_point(alpha=0.1) +
  ylim(0, 50) +
  xlim(0, 50) +
  labs(
    x = 'CpG count per 300 bp bin',
    y = 'Coverage of 300 bp bin'
  )

```

```{r medestrand_data}
medestrand_data <- tibble(abs_methylation = read_lines('~/../Downloads/medestrand.binmethyl.txt') %>% as.numeric)

medestrand_data %>%
  ggplot(aes(
    x = abs_methylation
  )) +
  geom_histogram()
```

```{r example_histograms_by_cpg, fig.width = 4, fig.height = 6}
chosen_cpg_count <- c(5, 12, 20)
chr_cpg_annotated %>%
  filter(
    cpg_count %in% chosen_cpg_count,
  ) %>%
  mutate(
    cpg_label = factor(
      sprintf('%s CpGs', cpg_count),
      levels = sprintf('%s CpGs', chosen_cpg_count)
    )) %>%
  ggplot(aes(
    x = mean_coverage
  )) +
  xlim(0, 50) +
  facet_grid(cpg_label ~ ., scales = 'free') +
  geom_histogram(bins = 50) +
  labs(
    x = 'Bin coverage',
    y = 'Count'
  )
```

```{r zero_profile_parameter_fit, fig.height = 3, fig.width = 5}
zero_profile_gc_model_output <- chr_cpg_annotated %>%
  filter(cpg_count == 0) %>%
  mutate(
    gc_bin = factor(round(gc_content * 4, 1) / 4)
  ) %>%
  group_by(gc_bin) %>%
  filter(n() > 50) %>%
  ungroup() %>%
  plyr::ddply(c('gc_bin'), function(z) {
    print(sprintf('Running for gc_bin = %s', z$gc_bin %>% unique))
    if (all(z$coverage_int == 0)) {
      return(tibble(
        log_mu = -Inf,
        theta = Inf,
        log_likelihood = NA,
        count = nrow(z)
      ))
    } else {
      tryCatch({
        zero_model <- flexmix(coverage_int ~ 1, data = z, k = 1, model = FLXMRnegbin())
        return(tibble(
          log_mu = parameters(zero_model, component=1)[[1]],
          theta = parameters(zero_model, component=1)[[2]],
          log_likelihood = logLik(zero_model) %>% as.numeric,
          count = nrow(z)
        ))
      }, error = function(e) {
        message('Error: skipping this GC bin')
        NULL
      })
    }
  }) %>%
  mutate(mu = exp(log_mu))

zero_mu_fit <- lm(
  log(mu) ~ log(gc_content),
  data = zero_profile_gc_model_output %>% mutate(gc_content = gc_bin %>% as.character %>% as.numeric) %>% filter(mu > 0)
)

zero_theta_fit <- lm(
  log(theta) ~ log(gc_content),
  data = zero_profile_gc_model_output %>% mutate(gc_content = gc_bin %>% as.character %>% as.numeric) %>% filter(!is.infinite(theta))
)

zero_profile_gc_model_output %>%
  mutate(gc_bin = as.numeric(as.character(gc_bin))) %>%
  dplyr::select(-log_mu) %>%
  gather(parameter, value, -count, -log_likelihood, -gc_bin) %>%
  filter(!is.infinite(value)) %>%
  ggplot(aes(
    x = gc_bin,
    y = value
  )) +
  geom_point(aes(
    color = parameter,
    group = parameter,
    size = count
  )) +
  geom_line(aes(
      x = gc_content,
      y = value,
      color = metric,
      group = metric
    ),
    data = tibble(
        gc_content = seq(0, 1, 0.01)
      ) %>%
      mutate(
        mu = exp(predict(zero_mu_fit, newdata = .)),
        theta = exp(predict(zero_theta_fit, newdata = .))
      ) %>%
      gather(metric, value, -gc_content)
  ) +
  scale_color_brewer(palette='Set1') +
  labs(
    x = 'GC Content',
    y = 'Parameter Value',
    size = 'Number of bins',
    color = 'Fit parameter'
  )
```

```{r zero_profile_simulated_data, fig.width = 6.5, fig.height = 4.5}
tibble(
  gc_content = seq(0, 0.8, 0.1)
) %>%
  mutate(
    mu = exp(predict(zero_mu_fit, newdata = .)),
    theta = exp(predict(zero_theta_fit, newdata = .))
  ) %>%
  plyr::ddply('gc_content', function(z) {
    tibble(x = 0:10) %>%
      mutate(y = dnbinom(x, z$theta, mu = z$mu))
  }) %>%
  ggplot(aes(
    x = x,
    y = y
  )) +
  facet_wrap(~ gc_content) +
  geom_bar(stat = 'identity') +
  ggtitle('Modeled non-specific binding by GC content (for CpG count = 0)') +
  labs(
    x = 'Coverage by bin',
    y = 'Relative frequency'
  )
```

```{r filter_out_zeros_method, fig.width = 30, fig.height=30}
# Create an initial estimate
bin_methylation <- chr_cpg_annotated %>%
  mutate(
    unmethylated_mu = exp(predict(zero_mu_fit, newdata=.)),
    unmethylated_theta = exp(predict(zero_theta_fit, newdata=.)),
    unmethylated_likelihood = dnbinom(coverage_int, mu = unmethylated_mu, size = unmethylated_theta),
    methylated_mu = ifelse(cpg_count == 0, NA, cpg_count),
    methylated_theta = ifelse(cpg_count == 0, NA, 0.5 * cpg_count),
    methylated_likelihood = ifelse(is.na(methylated_mu), 0, dnbinom(coverage_int, mu = methylated_mu, size = methylated_theta)),
    unmethylated_posterior = unmethylated_likelihood / (methylated_likelihood + unmethylated_likelihood),
    methylated_posterior = methylated_likelihood / (methylated_likelihood + unmethylated_likelihood)
  ) %>%
  mutate(methylation_status = ifelse(methylated_posterior > unmethylated_posterior, 'methylated', 'unmethylated'))

methylated_fits <- list()
MAX_ITER = 20
for (i in 1:MAX_ITER) {
  message(sprintf('Running iteration %s', i))
  bin_methylation_subset <- bin_methylation %>%
    filter(methylated_posterior > 0.5)
  
  # Fit NB regression to the methylated cases
  methylated_bins_nbfit <- glm.nb(
    coverage_int ~ cpg_count * gc_content,
    data = bin_methylation_subset
  )
  methylated_fits[[i]] = methylated_bins_nbfit
  
  # Refit the points based on the regression
  refitted <- chr_cpg_annotated %>%
    mutate(
      unmethylated_mu = exp(predict(zero_mu_fit, newdata=.)),
    unmethylated_theta = exp(predict(zero_theta_fit, newdata=.)),
      unmethylated_likelihood = dnbinom(coverage_int, mu = unmethylated_mu, size = unmethylated_theta),
      methylated_mu = ifelse(cpg_count == 0, NA, exp(predict(methylated_bins_nbfit, newdata = .))),
      methylated_theta = methylated_bins_nbfit$theta,
      methylated_likelihood = ifelse(is.na(methylated_mu), 0, dnbinom(coverage_int, mu = methylated_mu, size = methylated_theta)),
      unmethylated_posterior = unmethylated_likelihood / (methylated_likelihood + unmethylated_likelihood),
      methylated_posterior = methylated_likelihood / (methylated_likelihood + unmethylated_likelihood)
    )
  
  if (any(is.nan(refitted$unmethylated_posterior)) || any(is.nan(refitted$methylated_posterior))) {
    nan_rows = refitted %>% filter(is.nan(unmethylated_posterior) | is.nan(methylated_posterior))
    warning(sprintf('%s bins yielded NaNs - these may be outliers and have been removed. They are printed below.', nrow(nan_rows)))
    print(nan_rows)
    
    refitted <- refitted %>% filter(!is.nan(methylated_posterior), !is.nan(unmethylated_posterior))
  }
  
  bin_methylation <- refitted %>%
    mutate(
      methylation_status = ifelse(methylated_posterior > unmethylated_posterior, 'methylated', 'unmethylated')
    )
  
  bin_methylation_subset <- bin_methylation %>% filter(methylation_status == 'methylated')
  
  if (i > 1) {
    percent_changes = abs((methylated_fits[[i]]$coefficients - methylated_fits[[i-1]]$coefficients) / methylated_fits[[i-1]]$coefficients) * 100
    message('Percent Changes:')
    message(sprintf('%s  %s  %s\n', names(percent_changes), signif(percent_changes, 3), ifelse(percent_changes < 0.1, 'converged', '--')))
    if (all(percent_changes < 0.1)) {
      message('All coefficients converged.')
      methylated_fit <- methylated_bins_nbfit
      break
    } else if (i == MAX_ITER) {
      message('Maximum iterations hit. Some coefficients did not converge.')
    }
  }
}

```

```{r posterior_profiles, fig.width = 10, fig.height = 6}
bin_methylation %>%
  arrange(cpg_bin) %>%
  mutate(cpg_label = factor(sprintf('%s CpGs', cpg_bin), levels = unique(sprintf('%s CpGs', cpg_bin)))) %>%
  ggplot(aes(
    x = methylated_posterior,
    fill = methylation_status
  )) +
  geom_histogram() +
  facet_wrap(~ cpg_label, scales='free_y') +
  labs(
    x = 'Posterior Probability of Methylation',
    y = 'Count',
    color = 'Methylation Status'
  )
```

```{r posterior_methylated_fraction, fig.width = 8, fig.height = 5}
bin_methylation %>%
  group_by(cpg_count) %>%
  summarise(
    methylated = sum(methylation_status == 'methylated'),
    unmethylated = sum(methylation_status == 'unmethylated')
  ) %>%
  ungroup() %>%
  gather(methylation_status, count, -cpg_count) %>%
  group_by(cpg_count) %>%
  mutate(
    methylation_status = factor(methylation_status, levels = c('unmethylated', 'methylated')),
    fraction = count / sum(count),
    count = as.double(count)
  ) %>%
  ungroup() %>%
  ggplot(aes(
    x = cpg_count,
    y = fraction,
    color = methylation_status,
    group = methylation_status
  )) +
  geom_point(aes(
    size = count
  )) +
  geom_line() +
  scale_size_continuous(trans='log2') +
  labs(
    x = 'CpG Count',
    y = 'Fraction of bins',
    color = 'Methylation Status',
    size = 'Number of bins'
  )
```

```{r parameter_convergence_plots}
plyr::ldply(methylated_fits, function(z) {z$coeff}) %>%
  mutate(iteration = row_number()) %>%
  gather(metric, value, -iteration) %>%
  arrange(metric, iteration) %>%
  group_by(metric) %>%
  mutate(fractional_change = c(NA, abs((value[2:n()] - value[1:(n()-1)]) / value[1:(n() - 1)]))) %>%
  ungroup() %>%
  ggplot(aes(
    x = iteration,
    y = value
  )) +
  geom_line() +
  facet_wrap(~metric, scales='free')
```

```{r fit_distributions, fig.height=12, fig.width=15}
plot_bin_methylation_fit(
  bin_methylation %>% 
    filter(
      gc_content > 0.3,
      gc_content < 0.8,
      cpg_bin %in% c(4, 11, 14, 18)
    ) %>%
    group_by(cpg_bin, gc_bin) %>%
    filter(n() > 200) %>%
    ungroup(),
  methylated_fit
)
```

```{r fit_scatterplo, fig.height = 10, fig.width = 10}
plot_scatter_bin_methylation(bin_methylation, methylated_fit)
```

```{r heatmap_mean_coverage_by_cpg_gc_methylation, fig.height=5, fig.width = 6}
bin_methylation %>%
  filter(!is.na(cpg_region)) %>%
  mutate(gc_content = round(gc_content, 2)) %>%
  group_by(cpg_count, gc_content, cpg_region, methylation_status) %>%
  summarise(mean_mean_coverage = mean(mean_coverage)) %>%
  ungroup() %>%
  ggplot(aes(
    x = gc_content,
    y = cpg_count,
    fill = log(mean_mean_coverage)
  )) +
  geom_tile() +
  scale_fill_viridis() +
  facet_grid(methylation_status ~ cpg_region) +
  labs(
    x = 'GC content',
    y = 'CpG per 300 bp bin',
    fill = 'Log mean coverage'
  )
```

```{r coverage_by_cpg_scatterplot_colored_by_methylation_status}
bin_methylation %>%
  filter(mean_coverage < 75, cpg_count < 60) %>%
  ggplot(aes(
    x = cpg_count,
    y = mean_coverage,
    color = methylation_status
  )) +
  geom_point(shape = 20, alpha=0.1) +
  labs(
    x = 'CpG count per 300 bp bin',
    y = 'Coverage per 300 bp bin',
    color = 'Methylation status'
  )
```
# Downsampled data

```{r}
pipeline_output_downsampled <- read_tsv('../data/manual/CMP-01-08-cfDNA-1_downsampled_1000.tsv')

pipeline_output_downsampled %>%
  filter(coverage < 75, cpg_count < 60) %>%
  group_by(cpg_count, coverage) %>%
  summarise(
    methylated_fraction = sum(methylation_status == 'methylated') / n(),
    count = n()
  ) %>%
  ggplot(aes(
    x = cpg_count,
    y = coverage,
    fill = methylated_fraction,
    alpha = count
  )) +
  geom_tile() +
  labs(
    x = 'CpG count per 300 bp bin',
    y = 'Coverage per 300 bp bin',
    fill = 'Methylation status',
    alpha = 'Number of bins'
  ) +
  scale_alpha_continuous(trans='log2')
```
